File: tasks/javascript/7kyu/PartsOfAList.js
// Parts of a list

// Write a function partlist that gives all the ways to divide a list (an array) of at least two elements into two non-empty parts.

// Each two non empty parts will be in a pair (or an array for languages without tuples or a structin C - C: see Examples test Cases - )
// Each part will be in a string
// Elements of a pair must be in the same order as in the original array.
// Examples of returns in different languages:

// a = ["az", "toto", "picaro", "zone", "kiwi"] -->
// [["az", "toto picaro zone kiwi"], ["az toto", "picaro zone kiwi"], ["az toto picaro", "zone kiwi"], ["az toto picaro zone", "kiwi"]]
// or
//  a = {"az", "toto", "picaro", "zone", "kiwi"} -->
// {{"az", "toto picaro zone kiwi"}, {"az toto", "picaro zone kiwi"}, {"az toto picaro", "zone kiwi"}, {"az toto picaro zone", "kiwi"}}
// or
// a = ["az", "toto", "picaro", "zone", "kiwi"] -->
// [("az", "toto picaro zone kiwi"), ("az toto", "picaro zone kiwi"), ("az toto picaro", "zone kiwi"), ("az toto picaro zone", "kiwi")]
// or
// a = [|"az", "toto", "picaro", "zone", "kiwi"|] -->
// [("az", "toto picaro zone kiwi"), ("az toto", "picaro zone kiwi"), ("az toto picaro", "zone kiwi"), ("az toto picaro zone", "kiwi")]
// or
// a = ["az", "toto", "picaro", "zone", "kiwi"] -->
// "(az, toto picaro zone kiwi)(az toto, picaro zone kiwi)(az toto picaro, zone kiwi)(az toto picaro zone, kiwi)"
// Note

// You can see other examples for each language in "Your test cases"

// Sample Tests:
// const {assert,config} = require('chai')
// config.truncateThreshold = 0

// function testing(actual, expected) {
//     assert.deepEqual(actual, expected);
// }

// describe("partlist",function() {
//   it("1st Basic tests", function() {
//     testing(partlist(["I", "wish", "I", "hadn't", "come"]),
//         [["I", "wish I hadn't come"], ["I wish", "I hadn't come"], ["I wish I", "hadn't come"], ["I wish I hadn't", "come"]])
//     testing(partlist(["cdIw", "tzIy", "xDu", "rThG"]),
//         [["cdIw", "tzIy xDu rThG"], ["cdIw tzIy", "xDu rThG"], ["cdIw tzIy xDu", "rThG"]])
//     testing(partlist(["vJQ", "anj", "mQDq", "sOZ"]),
//         [["vJQ", "anj mQDq sOZ"], ["vJQ anj", "mQDq sOZ"], ["vJQ anj mQDq", "sOZ"]])
//   })
// })

function partlist(arr) {
  const result = [];

  for (let i = 1; i < arr.length; i++) {
    const part1 = arr.slice(0, i).join(" ");
    const part2 = arr.slice(i).join(" ");
    result.push([part1, part2]);
  }

  return result;
}

File: tasks/cpp/7kyu/Parts_of_a_list.cpp
/*
Write a function partlist that gives all the ways to divide a list (an array) of at least two elements into two non-empty parts.

Each two non empty parts will be in a pair (or an array for languages without tuples or a structin C - C: see Examples test Cases - )
Each part will be in a string
Elements of a pair must be in the same order as in the original array.
Examples of returns in different languages:
a = ["az", "toto", "picaro", "zone", "kiwi"] -->
[["az", "toto picaro zone kiwi"], ["az toto", "picaro zone kiwi"], ["az toto picaro", "zone kiwi"], ["az toto picaro zone", "kiwi"]] 
#include <vector>
#include <utility>

void testequal(std::vector<std::pair <std::string, std::string>> ans, std::vector<std::pair <std::string, std::string>> sol) {
    Assert::That(ans, Equals(sol));
}

Describe(partline_Tests)
{
    It(Fixed_Tests_partlist)
    {
        std::vector<std::string> s = {"cdIw", "tzIy", "xDu", "rThG"};
        std::vector<std::pair <std::string, std::string>> sol = {{"cdIw", "tzIy xDu rThG"}, {"cdIw tzIy", "xDu rThG"}, {"cdIw tzIy xDu", "rThG"}};
        std::vector<std::pair <std::string, std::string>> ans = PartList::partlist(s);
        testequal(ans, sol);

        s = {"I", "wish", "I", "hadn't", "come"};
        sol = {{"I", "wish I hadn't come"}, {"I wish", "I hadn't come"}, {"I wish I", "hadn't come"}, {"I wish I hadn't", "come"}};
        ans = PartList::partlist(s);
        testequal(ans, sol);

        s = {"vJQ", "anj", "mQDq", "sOZ"};
        sol = {{"vJQ", "anj mQDq sOZ"}, {"vJQ anj", "mQDq sOZ"}, {"vJQ anj mQDq", "sOZ"}};
        ans = PartList::partlist(s);
        testequal(ans, sol);
    }
};

  */
class PartList
{
public:
    static std::vector<std::pair <std::string, std::string>> partlist(std::vector<std::string> &arr) {
      std::vector<std::pair <std::string, std::string>> answ;
      std::string first = "";
      bool is_f = true;
      for (size_t i = 0; i < arr.size() - 1; ++i) {
        first += is_f ? arr[i] :" " + arr[i];
        is_f = false;
        std::string second = arr[i + 1];
        for (size_t j = i + 2; j < arr.size(); ++j) {
          second += " " + arr[j];
        }
        answ.push_back({first, second});
      }
      return answ;
    }
};
